src/shogun/mathematics/linalg/internal/implementation/MeanEigen3.h

```diff
@@ -0,0 +1,139 @@
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#ifndef MEAN_EIGEN_H_
+#define MEAN_EIGEN_H_
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/mathematics/linalg/internal/implementation/Sum.h>
+
+#include <shogun/mathematics/eigen3.h>
+#include <numeric>
+
+#include <iostream>
+
+namespace shogun
+{
+
+namespace linalg
+{
+    
+namespace implementation
+{
+    
+/**
+ * @brief Generic class int2float which converts different types of integer 
+ * into float64 type.
+ */
+template <typename inputType>
+struct int2float
+{
+    using floatType = inputType;
+};
+
+/**
+ * @brief Specialization of generic class int2float which converts int32 into float64.
+ */
+template <>
+struct int2float<int32_t>
+{
+    using floatType = float64_t;
+};
+
+/**
+ * @brief Specialization of generic class int2float which converts int64 into float64.
+ */
+template <>
+struct int2float<int64_t>
+{
+    using floatType = float64_t;
+};
+ 
+/**
+ * @brief Generic class mean which provides a static compute method. This class
+ * is specialized for different types of vectors and matrices, providing a mean
+ * to deal with various vectors and matrices directly.
+ */
+
+template <enum Backend, class Matrix>
+struct mean
+{
+	/** Scalar type */
+        typedef typename Matrix::Scalar T;
+        
+        /** int2float type */
+        typedef typename int2float<T>::floatType ReturnType;
+
+	/**
+	 * Method that computes the vector/matrix mean
+	 *
+	 * @param a vector/matrix whose mean has to be computed
+	 * @return the vector/matrix mean \f$\mean_i a_i\f$
+	 */
+        static ReturnType compute(Matrix a, bool no_diag=false);
+};
+
+/**
+ * @brief Specialization of generic mean which works with SGVector and 
+ * SGMatrix and uses Eigen3 as backend for computing mean.
+ */
+template <class Matrix>
+struct mean<Backend::EIGEN3, Matrix>
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+        
+        /** int2float type */
+        typedef typename int2float<T>::floatType ReturnType;
+        
+        /**
+	 * Method that computes the mean of SGVectors using Eigen3
+	 *
+	 * @param a vector whose mean has to be computed
+	 * @return the vector mean \f$\mean_i a_i\f$
+	 */
+        static ReturnType compute(SGVector<T> vec)
+	{
+		REQUIRE(vec.vlen > 0, "Vector cannot be empty!\n");
+                return (vector_sum<Backend::EIGEN3, SGVector<T> >::compute(vec)
+                 / ReturnType(vec.vlen));
```

Comment:
minor: we usually use tabs in Shogun, not spaces. Not too severe, but definitely should be udpated at some point